File: src/api/restaurants.ts
import api from "./client"

export async function getRestaurants() {
  const response = await api.get("/restaurants")
  return response.data
}

export async function getRestaurantById(id: string | number) {
  const response = await api.get(`/restaurants/${id}`)
  return response.data
}

export async function getRestaurantByUserId(userId: number) {
  const response = await api.get(`/restaurants/user/${userId}`)
  return response.data
}



export async function createRestaurant(data: {
  user_id: number
  name: string
  owner_name?: string
  description?: string
  image?: File | string | null
  phone?: string

  address_street?: string
  address_number?: string
  address_neighborhood?: string
  address_city?: string
  address_state?: string
  address_cep?: string

  latitude?: number | string | null
  longitude?: number | string | null

  delivery_fee?: number | string
  pix_key?: string
  bank_name?: string
  account_type?: string
  agency?: string
  account_number?: string
  document_number?: string
}) {
  const formData = new FormData()

  formData.append("user_id", String(Number(data.user_id)))
  formData.append("name", data.name.trim())

  if (data.owner_name?.trim()) formData.append("owner_name", data.owner_name.trim())
  if (data.description?.trim()) formData.append("description", data.description.trim())
  if (data.phone?.trim()) formData.append("phone", data.phone.trim())

  if (data.address_street?.trim()) formData.append("address_street", data.address_street.trim())
  if (data.address_number?.trim()) formData.append("address_number", data.address_number.trim())
  if (data.address_neighborhood?.trim()) formData.append("address_neighborhood", data.address_neighborhood.trim())
  if (data.address_city?.trim()) formData.append("address_city", data.address_city.trim())
  if (data.address_state?.trim()) formData.append("address_state", data.address_state.trim())
  if (data.address_cep?.trim()) formData.append("address_cep", data.address_cep.trim())

  if (data.latitude !== undefined && data.latitude !== null && data.latitude !== "") {
    formData.append("latitude", String(Number(data.latitude)))
  }

  if (data.longitude !== undefined && data.longitude !== null && data.longitude !== "") {
    formData.append("longitude", String(Number(data.longitude)))
  }

  formData.append("delivery_fee", String(Number(data.delivery_fee || 0)))

  if (data.pix_key?.trim()) formData.append("pix_key", data.pix_key.trim())
  if (data.bank_name?.trim()) formData.append("bank_name", data.bank_name.trim())
  if (data.account_type?.trim()) formData.append("account_type", data.account_type.trim())
  if (data.agency?.trim()) formData.append("agency", data.agency.trim())
  if (data.account_number?.trim()) formData.append("account_number", data.account_number.trim())
  if (data.document_number?.trim()) formData.append("document_number", data.document_number.trim())

  if (data.image instanceof File) {
    formData.append("image", data.image)
  }

  const response = await api.post("/restaurants", formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  })

  return response.data
}

File: app/api/routes/restaurants.py
from pathlib import Path
from uuid import uuid4

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.restaurant import Restaurant

router = APIRouter(prefix="/restaurants", tags=["Restaurants"])

UPLOAD_DIR = Path("uploads/restaurants")
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)


def save_uploaded_image(image: UploadFile | None) -> str | None:
    if image is None:
        return None

    if not image.content_type or not image.content_type.startswith("image/"):
        raise HTTPException(status_code=400, detail="O arquivo enviado deve ser uma imagem")

    extension = Path(image.filename).suffix.lower() if image.filename else ".jpg"
    filename = f"{uuid4().hex}{extension}"
    file_path = UPLOAD_DIR / filename

    with open(file_path, "wb") as f:
        f.write(image.file.read())

    return f"/uploads/restaurants/{filename}"


def serialize_restaurant(restaurant: Restaurant):
    return {
        "id": restaurant.id,
        "user_id": restaurant.user_id,
        "name": restaurant.name,
        "owner_name": restaurant.owner_name,
        "description": restaurant.description,
        "image": restaurant.image,
        "phone": restaurant.phone,
        "category": restaurant.category,
        "delivery_fee": float(restaurant.delivery_fee or 0),
        "address": {
            "street": restaurant.address_street,
            "number": restaurant.address_number,
            "neighborhood": restaurant.address_neighborhood,
            "city": restaurant.address_city,
            "state": restaurant.address_state,
            "cep": restaurant.address_cep,
        },
        "latitude": float(restaurant.latitude) if restaurant.latitude is not None else None,
        "longitude": float(restaurant.longitude) if restaurant.longitude is not None else None,
        "pix_key": restaurant.pix_key,
        "bank_name": restaurant.bank_name,
        "account_type": restaurant.account_type,
        "agency": restaurant.agency,
        "account_number": restaurant.account_number,
        "document_number": restaurant.document_number,
        "stripe_account_id": restaurant.stripe_account_id,
        "stripe_onboarding_complete": restaurant.stripe_onboarding_complete,

        # ⭐ ESTRELAS (ADICIONAR AQUI)
        "rating_average": float(restaurant.rating_average or 0),
        "rating_count": restaurant.rating_count or 0,
    }


@router.post("", status_code=status.HTTP_201_CREATED)
def create_restaurant(
    user_id: int = Form(...),
    name: str = Form(...),
    owner_name: str | None = Form(None),
    description: str | None = Form(None),
    phone: str | None = Form(None),
    category: str | None = Form(None),
    address_street: str | None = Form(None),
    address_number: str | None = Form(None),
    address_neighborhood: str | None = Form(None),
    address_city: str | None = Form(None),
    address_state: str | None = Form(None),
    address_cep: str | None = Form(None),
    latitude: float | None = Form(None),
    longitude: float | None = Form(None),
    delivery_fee: float = Form(0),
    pix_key: str | None = Form(None),
    bank_name: str | None = Form(None),
    account_type: str | None = Form(None),
    agency: str | None = Form(None),
    account_number: str | None = Form(None),
    document_number: str | None = Form(None),
    image: UploadFile | None = File(None),
    db: Session = Depends(get_db),
):
    image_url = save_uploaded_image(image)

    restaurant = Restaurant(
        user_id=user_id,
        name=name,
        owner_name=owner_name,
        description=description,
        image=image_url,
        phone=phone,
        category=category,
        address_street=address_street,
        address_number=address_number,
        address_neighborhood=address_neighborhood,
        address_city=address_city,
        address_state=address_state,
        address_cep=address_cep,
        latitude=latitude,
        longitude=longitude,
        delivery_fee=delivery_fee,
        pix_key=pix_key,
        bank_name=bank_name,
        account_type=account_type,
        agency=agency,
        account_number=account_number,
        document_number=document_number,
    )
    db.add(restaurant)
    db.commit()
    db.refresh(restaurant)
    return serialize_restaurant(restaurant)


@router.get("")
def list_restaurants(db: Session = Depends(get_db)):
    print("=== LISTANDO RESTAURANTES ===")
    restaurants = db.execute(select(Restaurant)).scalars().all()
    print("TOTAL:", len(restaurants))
    return [serialize_restaurant(r) for r in restaurants]


@router.get("/user/{user_id}")
def get_restaurant_by_user(user_id: int, db: Session = Depends(get_db)):
    restaurant = db.execute(
        select(Restaurant).where(Restaurant.user_id == user_id)
    ).scalar_one_or_none()

    if not restaurant:
        raise HTTPException(status_code=404, detail="Usuário não possui restaurante")

    return serialize_restaurant(restaurant)


@router.get("/{restaurant_id}")
def get_restaurant(restaurant_id: int, db: Session = Depends(get_db)):
    restaurant = db.get(Restaurant, restaurant_id)
    if not restaurant:
        raise HTTPException(status_code=404, detail="Restaurante não encontrado")
    return serialize_restaurant(restaurant)